src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java

```diff
@@ -127,6 +127,7 @@ private static boolean jmapHelp() {
         System.out.println("    --heap                  To print java heap summary.");
         System.out.println("    --binaryheap            To dump java heap in hprof binary format.");
         System.out.println("    --dumpfile <name>       The name of the dump file.");
+        System.out.println("    --gz <1-9>              The compression level for gzipped dump file.");
```

Comment:
Thanks for the thorough list, I have tested most of them but not all. I will cover them and update here later.

> So really there are two implementations of heap dumping, one in the VM and one in SA, but a total of 5 ways to get to them. I'm just starting to get my head wrapped around all this myself. We've created a very confusing situation for us, and unfortunately I think some of the changes that got us here was driven by a goal to simply things.

Yes, it makes me even curious about why there can be two different implementation of them (one by C++ and one by Java),  it seems we need double the work of adding new options for those tools, like parallal for jmap histo and dump, etc.

Reply:
One implementation is in the JVM itself, to be used when the JVM is still running well, and not just from command line tools. Heap dumping can also be triggered by the JVM itslef by setting flags like -XX:+HeapDumpBeforeFullGC. The other implementation is in SA, to be used on a core file or a hung JVM process, although it can also be used on a JVM that is still running well. BTW, there used to be a 3rd implementation. The old hprof profiler (a JVMTI agent) also was used to generate hprof files.

Reply:
Hi Lin,

A week ago you replied that you are accepting the following suggested refactoring:

```
                   /*
                    * Possible commands:
                    *     dumpheap gz=1 file
                    *     dumpheap gz=1
                    *     dumpheap file
                    *     dumpheap
                    *
                    * Use default filename if cntTokens == 0.
                    * Handle cases with cntTokens == 1 or 2.
                    */

                    if (cntTokens > 2) {
                        err.println("Too big number of options: " + cntTokens);
                        usage();
                        return;
                    }
                    if (cntTokens >= 1) { // parse first argument which is "gz=" option
                        String option  = t.nextToken();
                        gzlevel = parseHeapDumpCompressionLevel(option);
                        if (gzlevel == 0) {
                            usage();
                            return;
                        }
                        filename = "heap.bin.gz";
                    }
                    if (cntTokens == 2) { // parse second argument which is filename
                        filename = t.nextToken();
                        if (filename.startsWith("gz=")) {
                            err.println("Filename should not start with \"gz=\": " + filename);
                            usage();
                            return;
                        }
                    }
```

But I still do not see it in the latest updates from you. It feels like there is some misunderstanding and confusion.
Could you, please, clear this up?

Thanks,
Serguei

Reply:
Also, this method can be refactored to something like this (the check for exactly one option argument is needed):
```
   private int parseHeapDumpCompressionLevel(String option) {
       int gzl = 0;
       String[] keyValue = option.split("=");
       assert keyValue[0].equals("gz") : "expected option to start from gz=";
       if (keyValue.length != 2) {
           err.println("Exactly one argument is expected for option \"gz\"");
           return 0;
       }
       String level = keyValue[1];
       try {
           gzl = Integer.parseInt(level);
       } catch (NumberFormatException e) {
           err.println("gz option value not an integer ("+level+")");
           return 0;
       }
       if (gzl < 1 || gzl > 9) {
           err.println("Compression level out of range (1-9): " + level);
           return 0;
       }
       return gzl;
   }
```

Reply:
Hi Serguei,

> But I still do not see it in the latest updates from you. It feels like there is some misunderstanding and confusion.
Thanks a lot for remainder! Sorry that I forgot to submit the change.

> Also, this method can be refactored to something like this (the check for exactly one option argument is needed):
The change looks nice, I will made the change and submit it in a minute.

BRs,
Lin

Reply:
Hi Chris, 

> It's confusing because both `jhsdb jmap` and the clhsdb `dumpheap` command use the JMap class, but neither support passing `format=x` to it. In order to get SA to dump with `format=x`, you would need to launch `sun/jvm/hotspot/tools/JMap` directly and pass in `-heap:format=x`. I'm not suggesting you fix that, but just something for us to be aware of.

I think new PR can be created for adding `format=x` to `clhsdb dumpheap` and  `jhsdb jmap`. And maybe new CSR is required?

Reply:
Dear @plummercj,

> `--dumpfile` and `--gz` can only be used with `--binaryheap`. That should be made clear in the help text.

I added indentation for these two sub-options.

> * SA's `jhsdb jmap --binaryheap` (which uses JMap.java with `-heap:format=b`)
> * SA's clhsdb `dumpheap` command (which also uses JMap.java with `-heap:format=b`)
> * executing  SA's `sun/jvm/hotspot/tools/JMap` class directly
> * `jmap -dump` command (Uses Attach API's `heapdump` command, which uses hotspot `HeapDump` class)
> * `GC.heap_dump` dcmd, which also uses the hotspot `HeapDumper` class
All 5 commands help message have been tested.

Reply:
Hi Lin,
Thank you for the update with requested changes.
Please, note, my suggestion was rough and still needs to be tested/checked for potential bugs. :)
I'll look at changes in other files as well.
Thanks,
Serguei

Reply:
> I think new PR can be created for adding `format=x` to `clhsdb dumpheap` and `jhsdb jmap`. And maybe new CSR is required?

You can do that if you'd like. I'm not sure how important format=x is.